### PR TITLE
[KOGITO-3938] Add option to run tests with Kogito cluster wide operator

### PR DESCRIPTION
**Pull request by sutaakar**
Signed-off-by: Karel Suta <ksuta@redhat.com>

https://issues.redhat.com/browse/KOGITO-3938

Many thanks for submitting your Pull Request :heart:! 

Please make sure your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: `[KOGITO-XYZ] Subject`
- [x] Pull Request contains a link to the JIRA issue
- [x] Pull Request contains a description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [ ] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

### CODE REVIEW

**Comment by sutaakar** [comment]
Thinking about what should we use as a default, should we test cluster wide operator by default? Right now it is set so.

**Comment by sutaakar** [comment]
Right now the cluster installation fail in case there is already installed operator in the `kogito-operator-system` namespace, to prevent unintentional testing with old operator or parallel execution of multiple test suites. It has a downside - in case you run with `keep_namespace=true` then you need to manually delete the Kogito operator before running the test again.

Optionally I could add a check making sure that if you run with `keep_namespace=true` then the check for existing operator is skipped - existing operator is reused. What is your opinion on that?

File: test/main_test.go [hunk 1]
```diff
@@ -281,3 +290,76 @@ func deleteClusterWideTestOperators() {
...
+
+func monitorOlmNamespace() {
+	monitorNamespace(config.GetOlmNamespace())
+}
+
+func monitorKogitoOperatorNamespace() {
+	monitorNamespace(kogitoClusterWideNamespace)
+}
+
+func monitorNamespace(namespace string) {
```
Comment (radtriste) [inline test/main_test.go:302]: shouldn't we have that one in framework package ?

File: test/main_test.go [hunk 2]
```diff
@@ -281,3 +290,76 @@ func deleteClusterWideTestOperators() {
...
+
+func stopOlmNamespaceMonitoring() {
+	stopNamespaceMonitoring(config.GetOlmNamespace())
+}
+
+func stopKogitoOperatorNamespaceMonitoring() {
+	stopNamespaceMonitoring(kogitoClusterWideNamespace)
+}
+
+func stopNamespaceMonitoring(namespace string) {
```
Comment (radtriste) [inline test/main_test.go:319]: same for that one in framework package ?

**Comment by sutaakar** [comment]
BDD PR check failed with `panic: Namespace kogito-operator-system used to deploy cluster wide Kogito operator exists, Kogito operator may be already installed. Please uninstall the operator first`
There is a cluster wide operator installed. @Kaitou786 Can you remove it?

File: test/main_test.go [hunk 3]
```diff
@@ -281,3 +290,76 @@ func deleteClusterWideTestOperators() {
...
+		return err
+	}
+
+	// Deploy the operator
+	created, err := framework.DeployClusterWideKogitoOperatorFromYaml(kogitoClusterWideNamespace)
+	clusterObjects = created
+	if err != nil {
+		return err
+	}
+
```
Comment (radtriste) [inline test/main_test.go:348]: here we are not making sure that the Kogito Operator is running ?
if we don't, that will lead to errors in all scenarios or ?

Comment (radtriste) [reply test/main_test.go:348]: and then we should not need to wait for operator if cluster wide
- https://github.com/kiegroup/kogito-cloud-operator/pull/694/files#diff-87ffc708dd3b79b9732e75109db3f4ff2066ebb4d9e1dfb4964d4625710f7674R112-R114
- https://github.com/kiegroup/kogito-cloud-operator/pull/694/files#diff-87ffc708dd3b79b9732e75109db3f4ff2066ebb4d9e1dfb4964d4625710f7674R61-R63

**Comment by Kaitou786** [comment]
@sutaakar I removed the namespace, though just one observation, the namespace didn't had any workload in it. There was no deployment object with kogito-operator or anything. Should the check be more like, if namespace is present check for the deployment if there is no deployment create it and proceed?

**Review by radtriste — COMMENTED** [review]
I ran into a problem waiting for kogito operator in namespacewhereas it is supposed to be cluster-wide
```
{"level":"info","T":"2020-12-10T18:32:50.678+0100","logger":"kogito-operator-system","msg":"Create namespace kogito-operator-system"}
{"level":"info","T":"2020-12-10T18:32:50.705+0100","logger":"kogito-operator-system","msg":"Deploy Operator from yaml files in ../deploy/"}
{"level":"info","T":"2020-12-10T18:32:51.032+0100","logger":"kogito-operator-system","msg":"Wait 6m0s for image pulling secret"}
{"level":"info","T":"2020-12-10T18:32:52.041+0100","logger":"kogito-operator-system","msg":"Execute command oc [get secrets -o name -n kogito-operator-system]"}
{"level":"info","T":"2020-12-10T18:32:53.123+0100","logger":"kogito-operator-system","msg":"secret/builder-dockercfg-qc8dw\nsecret/builder-token-sghrv\nsecret/builder-token-tn6fq\nsecret/default-dockercfg-xk6qr\nsecret/default-token-4ccsp\nsecret/default-token-wr6sb\nsecret/deployer-dockercfg-wmpwx\nsecret/deployer-token-mq5xm\nsecret/deployer-token-q8rq2\nsecret/kogito-operator-dockercfg-llzjd\nsecret/kogito-operator-token-gkzl6\nsecret/kogito-operator-token-zxpgn\n"}
{"level":"info","T":"2020-12-10T18:32:53.124+0100","logger":"kogito-operator-system","msg":"'image pulling secret' is successful"}
{"level":"info","T":"2020-12-10T18:32:53.217+0100","logger":"tradisso-local-cucumber-da05","msg":"Scenario Install Kogito Data Index"}
{"level":"info","T":"2020-12-10T18:32:53.218+0100","logger":"tradisso-local-cucumber-da05","msg":"Step Namespace is created"}
{"level":"info","T":"2020-12-10T18:32:53.219+0100","logger":"tradisso-local-cucumber-da05","msg":"Create namespace tradisso-local-cucumber-da05"}
{"level":"info","T":"2020-12-10T18:32:53.240+0100","logger":"tradisso-local-cucumber-da05","msg":"Step Kogito Operator is deployed with Infinispan and Kafka operators"}
{"level":"info","T":"2020-12-10T18:32:53.270+0100","logger":"tradisso-local-cucumber-da05","msg":"Kogito operator deployment skipped as cluster wide operator should be deployed"}
{"level":"info","T":"2020-12-10T18:32:53.306+0100","logger":"tradisso-local-cucumber-da05","msg":"Subscribing to infinispan operator from source community-operators on channel 2.0.x"}
{"level":"info","T":"2020-12-10T18:32:53.566+0100","logger":"tradisso-local-cucumber-da05","msg":"Wait 30m0s for infinispan operator running"}
{"level":"info","T":"2020-12-10T18:33:56.590+0100","logger":"tradisso-local-cucumber-da05","msg":"'infinispan operator running' is successful"}
{"level":"info","T":"2020-12-10T18:33:56.590+0100","logger":"tradisso-local-cucumber-da05","msg":"Subscribing to strimzi-kafka-operator operator from source community-operators on channel stable in namespace openshift-operators"}
{"level":"info","T":"2020-12-10T18:33:56.618+0100","logger":"tradisso-local-cucumber-da05","msg":"Wait 30m0s for strimzi-kafka-operator operator in namespace openshift-operators running"}
{"level":"info","T":"2020-12-10T18:34:54.646+0100","logger":"tradisso-local-cucumber-da05","msg":"'strimzi-kafka-operator operator in namespace openshift-operators running' is successful"}
{"level":"info","T":"2020-12-10T18:34:54.646+0100","logger":"tradisso-local-cucumber-da05","msg":"Wait 15m0s for Kogito operator running"}
```

**Comment by radtriste** [comment]
> Thinking about what should we use as a default, should we test cluster wide operator by default? Right now it is set so.

In Nightlies, I would let the tests cluster-wide and we should maybe execute a subset (`@smoke` or another tag) with namespace operator.
In PR checks, lets let it cluster-wide, it should speedup a bit the testing. Maybe one simple runtime check with namespaced operator ?

**Comment by sutaakar** [comment]
> I ran into a problem waiting for kogito operator in namespacewhereas it is supposed to be cluster-wide

Good catch, I have missed that, will fix.

**Comment by sutaakar** [comment]
> Maybe one simple runtime check with namespaced operator ?

Not sure if it is worth it for PR check. For nightlies we can run smokes with namespace scoped.

File: test/main_test.go (hunk 1 above)
Comment (sutaakar) [reply test/main_test.go:302]: Will directly call `StartPodLogCollector`

File: test/main_test.go (hunk 2 above)
Comment (sutaakar) [reply test/main_test.go:319]: Not sure if it is worth it. This way it describes well the steps taken. Similar code is used in data.go too, though there are more steps defined.

File: test/main_test.go (hunk 3 above)
Comment (sutaakar) [reply test/main_test.go:348]: Will add the waiting code. The links above don't work :/

**Comment by radtriste** [comment]
> Right now the cluster installation fail in case there is already installed operator in the `kogito-operator-system` namespace, to prevent unintentional testing with old operator or parallel execution of multiple test suites. It has a downside - in case you run with `keep_namespace=true` then you need to manually delete the Kogito operator before running the test again.
> 
> Optionally I could add a check making sure that if you run with `keep_namespace=true` then the check for existing operator is skipped - existing operator is reused. What is your opinion on that?

`keep_namespace` was made to keep created components to be able to dig deeper if failing, so after the tests are run. 
I would not use it to "keep current operator if existing".
Usually that option is used for local testing or ? we should use it only on CRC maybe, not on OCP clusters IMO...

Also, thinking that for PR checks, we have a problem if run concurrently on the same cluster or ?
If we enable cluster wide operator BDD tests, we need to  [`disableConcurrentBuilds`](https://www.jenkins.io/doc/book/pipeline/syntax/#options) in pipelines. That could happen also for nightlies if release branch and master are run... until we have CRC ?

File: test/main_test.go (hunk 2 above)
Comment (radtriste) [reply test/main_test.go:319]: oki

Comment (radtriste) [reply test/main_test.go:319]: in that case, I would keep the `func monitorNamespace(namespace string) {`

File: test/main_test.go (hunk 3 above)
Comment (radtriste) [reply test/main_test.go:348]: I was meaning:
- https://github.com/kiegroup/kogito-cloud-operator/blob/master/test/steps/operator.go#L54-L56
- https://github.com/kiegroup/kogito-cloud-operator/blob/master/test/steps/operator.go#L100-L102
that goes with my comment on problems => https://github.com/kiegroup/kogito-cloud-operator/pull/694#issuecomment-742684258

**Comment by sutaakar** [comment]
>  keep_namespace was made to keep created components to be able to dig deeper if failing, so after the tests are run.

I use it also to quickly spin up custom operator image. Or possibly to experiment with deployed scenario. Anyway it is just for local or experimental usage. User needs to be aware of its limitations.

> Also, thinking that for PR checks, we have a problem if run concurrently on the same cluster or ?

Yes, that can be an issue. 

> If we enable cluster wide operator BDD tests, we need to disableConcurrentBuilds in pipelines. That could happen also for nightlies if release branch and master are run... until we have CRC ?

Right, good point.

File: test/main_test.go (hunk 3 above)
Comment (sutaakar) [reply test/main_test.go:353]: Now the Kogito operator waiting in scenario steps should happen only for namespace scoped deployment.

**Comment by sutaakar** [comment]
@radtriste The comments should be addressed now, please take a look again.
@Kaitou786 The Kogito operator deployment will now fail just in case there is already previous deployment in the namespace, in case of empty namespace it is reused.

**Review by radtriste — APPROVED** [review]
lgtm.
Could you also add the [`disableConcurrentBuilds`](https://www.jenkins.io/doc/book/pipeline/syntax/#options) option in `Jenkinsfile` and `Jenkinsfile.nightly` ?

**Comment by sutaakar** [comment]
@radtriste Done

**Comment by radtriste** [comment]
Waiting for BDD tests to get through and will merge.
I think the pipelines could be tweaked with milestones so that only the tests part is waiting to be run but that can be done in another PR ;)

**Comment by radtriste** [comment]
> I think the pipelines could be tweaked with milestones so that only the tests part is waiting to be run but that can be done in another PR ;)

Created https://issues.redhat.com/browse/KOGITO-4004

**Comment by sutaakar** [comment]
Green :)

**Pull request merged**